phys2bids/physio_obj.py

```diff
@@ -0,0 +1,403 @@
+    ------
+    var: any type
+        Variable to be checked (same as input).
+    """
+    if len(var) > data_size:
+        var = var[:data_size]
+
+    if len(var) < data_size:
+        is_valid(token, type(var))
+        var = var + [token] * (data_size - len(var))
+
+    return var
+
+
+class blueprint_input():
+    """
+    Main input object for phys2bids.
+    Contains the blueprint to be populated.
+    !!! Pay attention: there's rules on how to populate this object.
+    See below ("Attention") !!!
+
+    Input (Properties)
+    ------------------
+    timeseries : (ch, [tps]) list
+        List of numpy 1d arrays - one for channel, plus one for time.
+        Time channel has to be the first, trigger the second.
+        Contains all the timeseries recorded.
+        Supports different frequencies!
+    freq : (ch) list of floats
+        List of floats - one per channel.
+        Contains all the frequencies of the recorded channel.
+        Support different frequencies!
+    ch_name : (ch) list of strings
+        List of names of the channels - can be the header of the columns
+        in the output files.
+    units : (ch) list of strings
+        List of the units of the channels.
+
+    Properties
+    ----------
+    ch_amount: int
+        Number of channels (ch).
+
+    Optional properties
+    -------------------
+    num_timepoints_found: int
+        Amount of timepoints found in the automatic count,
+        *if* check_trigger_amount() is run
+
+    Methods
+    -------
+    rename_channels:
+        Changes the list "ch_name" in a controlled way.
+    return_index:
+        Returns the proper list entry of all the
+        properties of the object, given an index.
+    delete_at_index:
+        Returns all the proper list entry of the
+        properties of the object, given an index.
+    check_trigger_amount:
+        Counts the amounts of triggers and corrects time offset
+        in "time" ndarray. Also adds property ch_amount.
+
+    Attention
+    ---------
+    The timeseries (and as a consequence, all the other properties)
+    should start with an entry for time and an entry for trigger.
+    Both should have the same length - hence same sampling. Meaning:
+    - timeseries[0] → ndarray representing time
+    - timeseries[1] → ndarray representing trigger
+    - timeseries[0].shape == timeseries[1].shape
+
+    As a consequence:
+    - freq[0] == freq[1]
+    - ch_name[0] = 'time'
+    - ch_name[1] = 'trigger'
+    - units[0] = 's'
+    - Actual number of channels (ANC) +1 <= ch_amount <= ANC +2
+    """
+    def __init__(self, timeseries, freq, ch_name, units):
+        self.timeseries = is_valid(timeseries, list, list_type=np.ndarray)
+        self.ch_amount = len(self.timeseries)
+        self.freq = has_size(is_valid(freq, list,
+                                      list_type=(int, float)),
+                             self.ch_amount, 0)
+        self.ch_name = has_size(ch_name, self.ch_amount, 'unknown')
```

Comment:
I believe that `ch_name` here is generally a list, in which case this will fail with:

```python-traceback
~/phys2bids/physio_obj.py in is_valid(var, var_type, list_type, return_var)
     32     """
     33     if not isinstance(var, var_type):
---> 34         raise AttributeError(f'The given variable is not a {var_type}')
     35 
     36     if var_type is list and list_type is not None:

AttributeError: The given variable is not a <class 'list'>
```

Either the final parameter here should be ['unknown'] or we need to update the `has_size` and `is_valid` functions to better handle these different parameters combos we're hitting!

Reply:
The exact code that is being passed (and failing) is:

```python
self.ch_name = has_size(['time', 'trigger'], 7, 'unknown')
```

if you want to recreate the error locally. I was using the `.acq` file in the `tests/data` directory!

Reply:
Sorry, bug that wasn't failing on my part due to a change to my variables I did while debugging.
Corrected now though.